lib/types.js

```diff
@@ -484,18 +485,30 @@ registerType('long', {
     encoding(0x55, bufferOperations('Int8')),
     encoding(0x81, {
       encoder: function(val, bufb) {
-        if (val instanceof Int64) {
-          bufb.appendBuffer(val.toBuffer(true));
-        } else if (typeof val === 'number') {
-          var absval = Math.abs(val);
-          if (absval < 0xFFFFFFFF) {
-            bufb.appendUInt32BE(val < 0 ? 0xFFFFFFFF : 0x0);
-            bufb.appendUInt32BE(val < 0 ? (0xFFFFFFFF - absval + 1) : absval);
-          } else {
-            bufb.appendBuffer((new Int64(val)).toBuffer());
-          }
+        if (val instanceof Int64) return bufb.appendBuffer(val.toBuffer(true));
+        if (val instanceof Buffer) return bufb.appendBuffer(val);
+        if (typeof val !== 'number' || !(val instanceof Number)) val = Number(val);
+        if (!Number.isFinite(val)) {
+          throw new errors.EncodingError(val, 'invalid number');
+        }
+
+        var abs = Math.abs(val);
+        var high = abs / MAX_UINT;
+        var low = abs % MAX_UINT;
+        if (val > 0) {
+          bufb.appendInt32BE(high);
+          bufb.appendInt32BE(low);
         } else {
-          throw new errors.EncodingError(val, 'Invalid encoding type for 64-bit value');
+          // need to write to a buffer in order to calculate the 2s complement
+          var data = new Buffer(8), carry = 1;
+          for (var i = 7; i >= 0; i--) {
+            var value = ((low & 0xff) ^ 0xff) + carry;
+            data[i] = value & 0xff;
+            low = (i === 4) ? high : low >>> 8;
```

Comment:
I don't think you should reuse low here, I'd rather see a new variable name (e.g. `var curbytes = low;` before the for loop).